python/cuspatial/cuspatial/utils/binpred_utils.py

```diff
@@ -423,3 +400,62 @@ def _multipoints_is_degenerate(geoseries):
     ) & (y1.reset_index(drop=True) == y2.reset_index(drop=True))
     result[sizes_mask] = is_degenerate.reset_index(drop=True)
     return result
+
+
+def _pli_features_rebuild_offsets(pli, features):
+    in_sizes = (
+        features.sizes if len(features) > 0 else _zero_series(len(pli[0]) - 1)
+    )
+    offsets = cudf.Series(pli[0])
+    offset_sizes = offsets[1:].reset_index(drop=True) - offsets[
+        :-1
+    ].reset_index(drop=True)
+    in_indices = offset_sizes.index.repeat(offset_sizes)
+    if len(in_indices) == 0:
+        return _zero_series(len(pli[0]))
+    # Just replacing the indices from the feature series with
+    # the corresponding indices that are specified by the offsets buffer.
```

Comment:
Is this the "geometry index" (the part id that the current feature belongs to)?

Reply:
I'm not sure what you are asking. See the updated docs perhaps? Consider the example now in `_pli_points_to_multipoints`:

```
    pli = (
        cudf.Series([0, 2, 4, 6]),
        cuspatial.GeoSeries([
            Point(0, 0),
            LineString([(1, 1), (2, 2)]),
            Point(3, 3),
            Point(4, 4),
            Point(5, 5),
            LineString([(6, 6), (7, 7)]),
        ])
    )
```
In this example, each pair of features in the GeoSeries (indices [0, 1], [2, 3], and [4, 5]) were produced by input feature pairs (lhs[0], rhs[0]), (lhs[1], rhs[1]), and (lhs[2], rhs[2]) respectively. The offsets buffer [0, 2, 4, 6] specifies this. However, I only want the points, so this function takes the points from the pli result (and their corresponding indices):
```
print(pli[1][pli[1].feature_type == Feature_Enum.POINT.value])
0    POINT (0.00000 0.00000)
2    POINT (3.00000 3.00000)
3    POINT (4.00000 4.00000)
4    POINT (5.00000 5.00000)
dtype: geometry
```
and creates a new offset buffer using only these values, specified by the old `offset` groupings:
```
[0, 2, 4, 6] -> [0, 1, 3, 4]
```

Reply:
Thanks this is clear now.